File: bench/equivalents/python/sum_refined.py
#!/usr/bin/env python3
# Python equivalent of bench/sum-refined.ir: tail-style recursion over an
# accumulator whose type is only known through an integer? refinement, plus
# a closure created per step.
#
# Invoked by bench/bench.tcl; see fib.py for the --runs/value/best_us
# protocol shared by all four equivalents.
import sys
import time


def sum_refined(n, acc):
    if not isinstance(acc, int):
        raise ValueError("not_an_integer")
    if n == 0:
        return acc
    step = lambda x: x + n
    return sum_refined(n - 1, step(acc))


def parse_args(argv):
    runs = 5
    positional = []
    i = 0
    while i < len(argv):
        if argv[i] == "--runs":
            i += 1
            runs = int(argv[i])
        else:
            positional.append(argv[i])
        i += 1
    return runs, positional


if __name__ == "__main__":
    runs, positional = parse_args(sys.argv[1:])
    n0 = int(positional[0]) if len(positional) > 0 else 400
    acc0 = int(positional[1]) if len(positional) > 1 else 0

    sum_refined(n0, acc0)  # untimed warmup
    best = None
    value = None
    for _ in range(runs):
        t0 = time.perf_counter()
        value = sum_refined(n0, acc0)
        dt = (time.perf_counter() - t0) * 1_000_000
        if best is None or dt < best:
            best = dt

    print(f"value: {value}")
    print(f"best_us: {best:.3f}")

File: bench/equivalents/rust/sum_refined.rs
// Rust equivalent of bench/sum-refined.ir: tail-style recursion over an
// accumulator whose type is only known through an integer? refinement,
// plus a closure created per step.
//
// Invoked by bench/bench.tcl; see fib.rs for the --runs/value/best_us
// protocol shared by all four equivalents. Rust's static typing makes the
// integer? refinement check trivially true for an i64 accumulator; it is
// omitted rather than faked as a runtime check.
use std::hint::black_box;
use std::time::Instant;

#[inline(never)]
fn sum_refined(n: i64, acc: i64) -> i64 {
    if n == 0 {
        return acc;
    }
    let step = |x: i64| x + n;
    // black_box on the accumulator breaks the induction-variable chain the
    // optimizer would otherwise use to reduce the whole recursion to a
    // closed-form Gauss sum, so each step actually runs.
    sum_refined(n - 1, black_box(step(acc)))
}

fn main() {
    let mut runs: u32 = 5;
    let mut positional = Vec::new();
    let mut args = std::env::args().skip(1);
    while let Some(arg) = args.next() {
        if arg == "--runs" {
            runs = args.next().expect("--runs needs a value").parse().expect("--runs must be an integer");
        } else {
            positional.push(arg);
        }
    }
    let n0: i64 = positional.get(0).map(|s| s.parse().expect("n must be an integer")).unwrap_or(400);
    let acc0: i64 = positional.get(1).map(|s| s.parse().expect("acc must be an integer")).unwrap_or(0);

    sum_refined(black_box(n0), black_box(acc0)); // untimed warmup
    let mut best = f64::MAX;
    let mut value = 0;
    for _ in 0..runs {
        let t0 = Instant::now();
        // black_box on the arguments stops the optimizer from noticing the
        // call is loop-invariant and hoisting it out after the first run.
        value = sum_refined(black_box(n0), black_box(acc0));
        let us = t0.elapsed().as_secs_f64() * 1_000_000.0;
        if us < best {
            best = us;
        }
    }

    println!("value: {}", value);
    println!("best_us: {:.3}", best);
}
